includes/tracking/class-data-events.php

```diff
@@ -0,0 +1,191 @@
+		);
+	}
+
+	/**
+	 * Returns whether a product is a one time purchase, or recurring and when.
+	 *
+	 * @param string $product_id Product's ID.
+	 */
+	public static function get_purchase_recurrence( $product_id ) {
+		$recurrence = get_post_meta( $product_id, '_subscription_period', true );
+		if ( empty( $recurrence ) ) {
+			$recurrence = 'once';
+		}
+		return $recurrence;
+	}
+
+	/**
+	 * Returns whether a product ID is associated with a membership.
+	 *
+	 * @param string $product_id Product's ID.
+	 */
+	public static function is_membership_product( $product_id ) {
+		if ( ! function_exists( 'wc_memberships_get_membership_plans' ) ) {
+			return false;
+		}
+		$membership_plans = wc_memberships_get_membership_plans();
+		$plans            = [];
+
+		foreach ( $membership_plans as $plan ) {
+			$subscription_plan  = new \WC_Memberships_Integration_Subscriptions_Membership_Plan( $plan->get_id() );
+			$required_products = $subscription_plan->get_subscription_product_ids();
+			if ( in_array( $product_id, $required_products ) ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+
+	/**
+	 * Returns the product type: product, subscription, donation, or membership.
+	 * TODOGA4: move this check & related functions into a more central location, and update based on final decision for product types.
+	 *
+	 * @param string $product_id Product's ID.
+	 */
+	public static function get_product_type( $product_id ) {
+		$product_type = 'product';
+		$recurrence   = self::get_purchase_recurrence( $product_id );
+
+		// Check if it's a subscription product.
+		if ( 'once' !== $recurrence ) {
+			$product_type = 'subscription';
+		}
+
+		// Check if it's a membership product.
+		if ( self::is_membership_product( $product_id ) ) {
+			$product_type = 'membership';
+		}
+
+		// Check if it's a donation product.
+		if ( method_exists( 'Newspack\Donations', 'is_donation_product' ) ) {
+			if ( \Newspack\Donations::is_donation_product( $product_id ) ) {
+				$product_type = 'donation';
+			}
+		}
+
+		return $product_type;
+	}
+
+	/**
+	 * Returns the action type: checkout_button or donation.
+	 *
+	 * @param string $product_id Product's ID.
+	 */
+	public static function get_action_type( $product_id ) {
+		$action_type = 'checkout_button';
+		// Check if it's a donation product, and update action_type, product_type.
+		if ( method_exists( 'Newspack\Donations', 'is_donation_product' ) ) {
+			if ( \Newspack\Donations::is_donation_product( $product_id ) ) {
+				$action_type = 'donation';
+			}
+		}
+		return $action_type;
+	}
+
+	/**
+	 * Returns an array of product information.
+	 *
+	 * @param string $product_id Product's ID.
+	 * @param array  $cart_item Cart item.
+	 * @param string $product_parent_id Product's ID when it has variations.
+	 */
+	public static function build_js_data_events( $product_id, $cart_item, $product_parent_id = '' ) {
+		// Reassign the IDs to make sure the product is the product and the variation is the variation.
```

Comment:
Assigning `$variation_id` with an empty string value helps avoid a `PHP Warning:  Undefined variable $variation_id in /newspack-repos/newspack-blocks/includes/tracking/class-data-events.php on line 154` when interacting with a non-variable product. We can also simplify the check of `$product_parent_id` slightly with just one `! empty()` check.

```suggestion
		// Reassign the IDs to make sure the product is the product and the variation is the variation.
		$variation_id = '';
		if ( ! empty( $product_parent_id ) ) {
			$variation_id = $product_id;
			$product_id   = $product_parent_id;
		}
```

Reply:
Ah, makes sense! Added 🙂